src/elevateprivilegeshelper.cpp

```diff
@@ -0,0 +1,186 @@
+#include "elevateprivilegeshelper.h"
+
+#include <QDebug>
+#include <QFile>
+#include <QSocketNotifier>
+#include <QStandardPaths>
+#include <QUuid>
+
+#include <fcntl.h>
+#include <signal.h>
+#include <sys/stat.h>
+#include <sys/types.h>
+#include <sys/wait.h>
+#include <vector>
+
+InitiallyStoppedProcess::~InitiallyStoppedProcess()
+{
+    kill();
+}
+
+bool InitiallyStoppedProcess::reset(const QString& exePath, const QStringList& exeOptions,
+                                    const QString& workingDirectory)
+{
+    kill();
+
+    // convert arguments and working dir into what the C API needs
+
+    std::vector<QByteArray> args;
+    args.reserve(exeOptions.size() + 1);
+    args.emplace_back(exePath.toLocal8Bit());
+    for (const auto& opt : exeOptions)
+        args.emplace_back(opt.toLocal8Bit());
+    const auto wd = workingDirectory.toLocal8Bit();
+
+    std::vector<char*> argsArray(args.size() + 1);
+    for (size_t i = 0; i < args.size(); ++i)
+        argsArray[i] = args[i].data();
+    argsArray.back() = nullptr;
+
+    // fork
+    m_pid = fork();
+
+    if (m_pid == 0) { // inside child process
+        // change working dir
+        if (!wd.isEmpty() && chdir(wd.data()) != 0)
+            qFatal("Failed to change working directory to %s", wd.data());
+
+        // stop self
+        raise(SIGSTOP);
+
+        // exec
+        execvp(argsArray[0], argsArray.data());
+        qFatal("Failed to exec %s", argsArray[0]);
+    } else if (m_pid < 0) {
+        qCritical("Failed to fork (?)");
+        return false;
+    }
+
+    return true;
+}
+
+bool InitiallyStoppedProcess::run()
+{
+    if (m_pid <= 0)
+        return false;
+
+    // wait for child to be stopped
+
+    int wstatus;
+    waitpid(m_pid, &wstatus, WUNTRACED);
+    if (!WIFSTOPPED(wstatus)) {
+        m_pid = -1;
+        return false;
+    }
+
+    // continue
+
+    ::kill(m_pid, SIGCONT);
+    return true;
+}
+
+void InitiallyStoppedProcess::terminate()
+{
+    if (m_pid > 0)
+        ::kill(m_pid, SIGTERM);
+}
+
+void InitiallyStoppedProcess::kill()
+{
+    if (m_pid > 0) {
+        ::kill(m_pid, SIGKILL);
+        waitpid(m_pid, nullptr, 0);
+        m_pid = -1;
+    }
+}
+
+PerfControlFifoWrapper::~PerfControlFifoWrapper()
+{
+    close();
+}
+
+bool PerfControlFifoWrapper::open()
+{
+    close();
+
+    QString fifoParentPath = QStandardPaths::writableLocation(QStandardPaths::RuntimeLocation);
+    if (fifoParentPath.isEmpty())
+        fifoParentPath = QStandardPaths::writableLocation(QStandardPaths::TempLocation);
+
+    const auto fifoBasePath =
+        QStringLiteral("%1/hotspot-%2-%3-perf")
+            .arg(fifoParentPath, QString::number(getpid()), QUuid::createUuid().toString().mid(1, 6));
+    m_ctlFifoPath = fifoBasePath + QStringLiteral("-control.fifo");
+    m_ackFifoPath = fifoBasePath + QStringLiteral("-ack.fifo");
+
+    if (mkfifo(m_ctlFifoPath.toLocal8Bit().data(), 0600) != 0) {
```

Comment:
DRY: can you introduce a reusable function that does the mkfifo + open and then call that once for the ack and once for the ctl fifo?

also note that you are converting the QString to QByteArray twice for each fifo